sys/net/application_layer/ng_zep/ng_zep.c

```diff
@@ -0,0 +1,1072 @@
+void *_event_loop(void *args);
+
+/* Builds (uninitialized) ZEP packet according to configured version */
+static ng_pktsnip_t *_zep_hdr_build(ng_zep_t *dev, size_t size, bool ack);
+
+/* Fills ZEP header according to ng_zep_t configuration */
+static size_t _zep_hdr_fill(ng_zep_t *dev, ng_zep_hdr_t *hdr,
+                            size_t payload_len);
+
+/* Event handlers for ISR events */
+static void _rx_started_event(ng_zep_t *dev);
+
+/* IEEE 802.15.4 helper functions: TODO: generalize add to ng_ieee802154 */
+static size_t _make_data_frame_hdr(ng_zep_t *dev, uint8_t *buf,
+                                   ng_netif_hdr_t *hdr);
+static size_t _get_frame_hdr_len(uint8_t *mhr);
+ng_pktsnip_t *_make_netif_hdr(uint8_t *mhr);
+static uint16_t _calc_fcs(uint16_t fcs, const uint8_t *frame, uint8_t frame_len);
+
+kernel_pid_t ng_zep_init(ng_zep_t *dev, uint16_t src_port, ng_ipv6_addr_t *dst,
+                         uint16_t dst_port)
+{
+#if CPUID_ID_LEN
+    uint8_t cpuid[CPUID_ID_LEN];
+    uint32_t hash1, hash2;
+#endif
+
+    if (_pid != KERNEL_PID_UNDEF) {
+        DEBUG("zep: ZEP thread already running at pid=%" PRIkernel_pid "\n", _pid);
+        return -EEXIST;
+    }
+
+    if (dev == NULL) {
+        DEBUG("zep: dev was NULL\n");
+        return -ENODEV;
+    }
+
+    if ((dst == NULL) || (ng_ipv6_addr_is_unspecified(dst))) {
+        DEBUG("zep: dst (%s) was NULL or unspecified\n", dst);
+        return -ENOTSUP;
+    }
+
+    if (ng_netreg_lookup(NG_NETTYPE_UDP, src_port)) {
+        DEBUG("zep: port (%" PRIu16 ") already registered\n", src_port);
+        return -EADDRINUSE;
+    }
+
+    dev->driver = (ng_netdev_driver_t *)&_zep_driver;
+    dev->chan = NG_ZEP_DEFAULT_CHANNEL;
+    dev->pan = byteorder_btols(byteorder_htons(NG_ZEP_DEFAULT_PANID));
+    dev->flags = NG_ZEP_FLAGS_USE_SRC_PAN;
+#if CPUID_ID_LEN
+    /* initialize dev->addr and dev->eui64 from cpuid if available */
+    cpuid_get(cpuid);
+
+    hash1 = djb2_hash(cpuid, CPUID_ID_LEN / 2);
+    dev->addr.u16 = (uint16_t)((hash1 >> 16) ^ (hash1 & 0xffff));
+
+    if (CPUID_ID_LEN % 2) {
+        hash2 = djb2_hash(cpuid + (CPUID_ID_LEN / 2), (CPUID_ID_LEN / 2) - 1);
+    }
+    else {
+        hash2 = djb2_hash(cpuid + (CPUID_ID_LEN / 2), CPUID_ID_LEN / 2);
+    }
+
+    dev->eui64.u32[0] = hash1;
+    dev->eui64.u32[1] = hash2;
+
+    dev->eui64.u8[7] &= 0xfe;   /* set to unicast */
+    dev->eui64.u8[7] |= 0x02;   /* set to locally administered */
+#else
+    dev->addr = NG_ZEP_DEFAULT_ADDR_SHORT;
+    dev->eui64 = NG_ZEP_DEFAULT_ADDR_LONG;
+#endif
+    DEBUG("zep: initialized radio parameters: chan: %" PRIu8 ", pan: 0x%04" PRIx16
+          "addr: 0x%04" PRIx16 ", eui64: %016" PRIx64 "\n", dev->chan,
+          dev->pan, byteorder_ltobs(dev->addr).u16,
+          byteorder_ltobll(dev->eui64).u64);
+
+#ifdef MODULE_NG_SIXLOWPAN
+    dev->proto = NG_NETTYPE_SIXLOWPAN;
+#else
+    dev->proto = NG_NETTYPE_UNDEF;
+#endif
+
+    dev->seq = genrand_uint32();
```

Comment:
For future usage this is definitely how it should be.

In the current configuration `random` is also included for `ng_ndp`, too. If we have another random-less network layer in the future we can discuss this further.